### ISSUE TITLE
System.IndexOutOfRangeException in Dune 2000

### ISSUE THREAD

**Issue by Haukness** [OpenRA/OpenRA#11333]
Good evening OpenRA developers!

When playing Dune 2000, the game often crashes, usually when in combat with enemy units.  Here is the exception.log from the most recent crash:

Dune 2000 Mod at Version release-20160508
on map b9c74c6501ba964e2eaf0eddc617b7734b126cda (The Duell by CombineCnR).
Operating System: Windows (Microsoft Windows NT 5.2.3790 Service Pack 2)
Runtime Version: .NET CLR 4.0.30319.1022
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.Traits.MPStartLocations.<WorldLoaded>c__AnonStorey0.<>m__0(Client c)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at OpenRA.Mods.Common.Traits.MPStartLocations.WorldLoaded(World world, WorldRenderer wr)
   at OpenRA.World.LoadComplete(WorldRenderer wr)
   at OpenRA.Game.StartGame(String mapUID, WorldType type)
   at OpenRA.Network.UnitOrders.ProcessOrder(OrderManager orderManager, World world, Int32 clientId, Order order)
   at OpenRA.Network.OrderManager.TickImmediate()
   at OpenRA.Sync.<CheckSyncUnchanged>c__AnonStorey0.<>m__0()
   at OpenRA.Sync.CheckSyncUnchanged[T](World world, Func`1 fn)
   at OpenRA.Sync.CheckSyncUnchanged(World world, Action fn)
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Program.Run(String[] args)
   at OpenRA.Program.Main(String[] args)

I'm not sure exactly how to reproduce the crash, but each time it has happened, my units have been in combat with enemy units.  The game freezes, the sound loops for several seconds, and then I receive the warning that the game has crashed.

The issue is intermittent, and apart from it, the game runs well.  Any help you guys can offer would be greatly appreciated!

**Comment by obrakmann** [comment]
This exception doesn't look like it matches the description of your crash, since it would have happened right at the start of a game, not in the middle of it.

**Comment by obrakmann** [comment]
This particular crash happens due to a bug in that map. The player spawns are owned by the players themselves, so when one selects the spawnpoint being owned by the other player - and that other player being absent so that the spawn is not actually added to the world - the game crashes at the start of the match.

**Comment by Haukness** [comment]
That's strange.  Almost every time I've encountered crashing problems with Dune 2000, it happens in the middle of a battle, usually about 30 minutes or so in.  There ARE some maps that cause the game to crash immediately after the battle starts, as you say, but this particular map isn't one of them.  I didn't check the exception.log file in these instances, as I figured there was just something wrong with the map I was trying to play on.

To provide a bit more detail, the map included in this particular crash allows for up to 8 players.  However, I only had six players active in this particular battle:  Myself, an AI ally (both of us on team 2), and then four enemy AIs not assigned to any teams.  The two northern-most spawn points on the map were empty.  When setting up the game, I left the last two slots "Open", rather than closing them, figuring it wouldn't make a difference.  Do you suppose closing the empty slots might prevent this crash on the same map?

Also, I tried playing on this map with a similar player configuration on a different computer, and the game crashed on that machine as well about 30 minutes into the battle.  However, the game did not placed an exception.log file in the Logs folder on that machine, so I wasn't able to determine if the error was the same, or different.

Do you have any suggestions on what I could do to prevent the problem, or how to reproduce it more reliably to allow for further troubleshooting?

I appreciate your time helping me with this!

**Comment by obrakmann** [comment]
I'm not saying that you don't experience those crashes, just that the exception you posted doesn't match your description (but another bug which I have then identified and fixed, so thanks for the report already :smiley:). You can see that your report and the crash log don't match by the map name, since you said it's happening on an 8-player map, but "The Duell" is a two-player map.

When you experience those other crashes mid-game, get the exception.log file immediately (and check its timestamp!) and open a new bug, please.

(You should also upgrade your OS, Windows Server 2003 has been out of support for almost 11 months.)

Thanks!

**Comment by obrakmann** [comment]
Just out of curiosity: what's your GPU?

**Comment by Haukness** [comment]
obrakmann,

No worries, I understood what you meant and were just trying to get to the bottom of the error!  I, too, noticed that the map name didn't quite match as well.

I'll keep playing the game and trying to replicate the crashes, and make sure to check the date on exception.log before opening a new bug report!

Also, the computer I sent you the log file from is actually running Windows XP Professional x64 (which is based on Server 2003), which has been out of support since April of 2014.  And yes, I'm definitely overdue for an upgrade!  I'm planning to construct a new machine soon, so I hope not to be using it very much longer.

It's worth mentioning that I have also replicated this "Crash after 30 minutes or so" issue on my laptop, which is running Windows 7 Professional x64.

To answer your question about my GPUs, each machine uses the following:

Desktop (WinXP x64): GeForce GTX 280 (Driver Version 6.14.13.681, Driver Date 9/23/2012)
Laptop (Win7 x64): GeForce 9800M GTS (Driver Version 9.18.13.623, Driver Date 8/30/2012)

Also, question.  Does the program write any debugging or error information to any of the other log files apart from exception.log?  After replicating the crash on my laptop, I looked in the Logs folder, but no exception.log file was present.  A few of the other logs had data in them, but I wasn't sure if any of them were relevant to my problem.

Once again, I really appreciate you taking the time to troubleshoot this.  I'll open a new bug report once I have a fresh copy of a log to provide you!
